fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -1438,52 +1439,62 @@ private RecordQueryPlan tryToConvertToCoveringPlan(@Nonnull PlanContext context,
+                                                                                 @Nullable KeyExpression commonPrimaryKey,
+                                                                                 @Nonnull Set<KeyExpression> requiredFields) {
+        final Collection<RecordType> recordTypes = metaData.recordTypesForIndex(index);
+        if (recordTypes.size() != 1) {
+            return null;
         }
+        final RecordType recordType = Iterables.getOnlyElement(recordTypes);
+
+        final KeyExpression rootExpression = index.getRootExpression();
+        final List<KeyExpression> keyFields = KeyExpression.getKeyFields(rootExpression);
+        final List<KeyExpression> valueFields = KeyExpression.getValueFields(rootExpression);
 
         // Like FDBRecordStoreBase.indexEntryKey(), but with key expressions instead of actual values.
-        final List<KeyExpression> primaryKeys = context.commonPrimaryKey == null
-                ? Collections.emptyList()
-                : context.commonPrimaryKey.normalizeKeyForPositions();
+        final List<KeyExpression> primaryKeys = commonPrimaryKey == null
+                                                ? Collections.emptyList()
+                                                : commonPrimaryKey.normalizeKeyForPositions();
         index.trimPrimaryKey(primaryKeys);
         keyFields.addAll(primaryKeys);
 
-        final IndexKeyValueToPartialRecord.Builder builder = IndexKeyValueToPartialRecord.newBuilder(recordType.getDescriptor());
+        final IndexKeyValueToPartialRecord.Builder builder = IndexKeyValueToPartialRecord.newBuilder(recordType);
 
-        for (KeyExpression resultField : resultFields) {
-            if (!addCoveringField(resultField, builder, keyFields, valueFields)) {
-                return chosenPlan;
-            }
+        Set<KeyExpression> fields = new HashSet<>(requiredFields);
+        if (commonPrimaryKey != null) {
+            // Need the primary key, even if it wasn't one of the explicit result fields.
+            fields.addAll(commonPrimaryKey.normalizeKeyForPositions());
         }
-
-        if (context.commonPrimaryKey != null) {
-            for (KeyExpression primaryKeyField : context.commonPrimaryKey.normalizeKeyForPositions()) {
-                // Need the primary key, even if it wasn't one of the explicit result fields.
-                if (!resultFields.contains(primaryKeyField)) {
-                    addCoveringField(primaryKeyField, builder, keyFields, valueFields);
-                }
+        for (KeyExpression field: fields) {
+            if (!addCoveringField(field, builder, keyFields, valueFields)) {
+                return null;
             }
         }
 
+        // Validity check ensures that we don't attempt this transformation if there are any repeated fields..
         if (!builder.isValid()) {
-            return chosenPlan;
+            return null;
```

Comment:
`chosenPlan` ?

Reply:
oh got it -- different contract in a different method

Reply:
`chosenPlan` is the index plan before the covering index transformation. I pulled out the building logic from the convert-to-covering logic so that it could be reused by the visitor. Returning `null` is the equivalent for the new control flow.